content/blog/2018-02-07-strict-mode.md

```diff
@@ -0,0 +1,69 @@
+
+When strict mode is enabled, React compiles a list of all class components using the unsafe lifecycles, and logs a warning message with information about these components, like so:
+
+![](../images/blog/strict-mode-unsafe-lifecycles-warning.png)
+
+Addressing the issues identified by strict mode _now_ will make it easier for you to take advantage of async rendering in future releases of React.
+
+### Warning about legacy string ref API usage
+
+Previously, React provided two ways for managing refs: the legacy string ref API and the callback API. Although the string ref API was the more convenient of the two, it had [several downsides](https://github.com/facebook/react/issues/1373) and so our official recomendation was to [use the callback form instead](https://reactjs.org/docs/refs-and-the-dom.html#legacy-api-string-refs).
+
+Version 16.3 adds a new option for managing refs that offers the convenience of a string ref without any of the downsides:
+`embed:16-3-release-blog-create-ref.js`
+
+> **Note:**
+>
+> Callback refs will continue to be supported in addition to the new `createRef` API.
+>
+> You don't need to replace callback refs in your components. They are slightly more flexible, so they will remain as an advanced feature.
+
+[Learn more about the new `createRef` API here.](#)
+
+### Detecting unexpected side effects
+
+As a general rule, side-effects should be avoided in certain class component methods (e.g. the `constructor`, `render`, etc). This is because React may invoke these methods more than once before committing, or it may invoke them without committing at all (because of an error or a higher priority interruption). Ignoring this rule can lead to a variety of problems, including memory leaks and invalid state. Unfortunately, it can be difficult to detect these problems as they are often non-deterministic.
+
+Strict mode can't automatically detect side effects for you, but it can help you spot them by making them a little more deterministic. This is done by intentionally double-invoking the following methods:
+
+* Class component `constructor` method
+* The `render` method
+* `setState` updater functions
+* The static `getDerivedStateFromProps` lifecycle
+
+> Note:
+>
+> This only applies to development mode. _Lifecycles will not be double-invoked in production mode._
+
+For example, consider the following code:
+`embed:update-on-async-rendering/side-effects-in-constructor.js`
+
+At first glance, this code might not seem problematic. But if `SharedApplicationState.recordEvent` is not idempotent, then instantiating this component multiple times could lead to invalid application state. This sort of subtle bug might not manifest during development, or it might do so inconsistently and so be overlooked.
```

Comment:
"idempotent" should be a link to wikipedia or some explanation. We shouldn't assume React users know these terms

Reply:
Hm. I'm willing to add a link, but I think it's not unreasonable to assume that someone can Wikipedia search for a fairly common programming term if they don't know the answer. (I have to do this all the time, myself.)

Reply:
I've done UI programming for years and never encountered this term. I agree people can look it up, but in practice people who don't understand it will likely either skip it or feel inadequate

Reply:
I've already added a link (locally).

I was just expressing my point of view: I have to look up terms that I'm uncertain about regularly. I just didn't think it was unreasonable not to have a link in this case.

Reply:
I think maybe I based the tone/tech-level of this blog post too heavily on recent ones (like the 16.0 announcement) which contain wording like so:

>  async rendering—a strategy for cooperatively scheduling rendering work by periodically yielding execution to the browser. The upshot is that, with async rendering, apps are more responsive because React avoids blocking the main thread.